File: src/lua/utf-256.lua
#!/usr/bin/env lua
local CHUNK_SIZE = 8

local function print_help(progname)
    io.stderr:write([[
UTF-256 Encoder/Decoder

Usage:
  ]] .. progname .. [[ -e UTF-8_FILE -o OUTPUT_FILE
  ]] .. progname .. [[ -d UTF-256_FILE -o OUTPUT_FILE

]])
end

local function encode(infile, outfile)
    local in_file = io.open(infile, "rb")
    if not in_file then
        io.stderr:write("Error opening input file\n")
        return 1
    end

    local out_file = io.open(outfile, "wb")
    if not out_file then
        io.stderr:write("Error opening output file\n")
        in_file:close()
        return 1
    end

    while true do
        local byte = in_file:read(1)
        if not byte then break end
        local b = string.byte(byte)
        for i = 7, 0, -1 do
            local bit = (b >> i) & 1
            local out_byte = bit == 1 and 0xFF or 0x00
            out_file:write(string.char(out_byte))
        end
    end

    in_file:close()
    out_file:close()
    return 0
end

local function decode(infile, outfile)
    local in_file = io.open(infile, "rb")
    if not in_file then
        io.stderr:write("Error opening input file\n")
        return 1
    end

    local out_file = io.open(outfile, "wb")
    if not out_file then
        io.stderr:write("Error opening output file\n")
        in_file:close()
        return 1
    end

    while true do
        local chunk = in_file:read(CHUNK_SIZE)
        if not chunk then break end
        if #chunk < CHUNK_SIZE then
            io.stderr:write("Error: file is not a multiple of 8 bytes\n")
            in_file:close()
            out_file:close()
            return 1
        end

        local byte = 0
        for i = 1, CHUNK_SIZE do
            local c = string.byte(chunk, i)
            if c == 0x00 then
                byte = (byte << 1)
            elseif c == 0xFF then
                byte = (byte << 1) | 1
            else
                io.stderr:write(string.format("Invalid byte in UTF-256 file: 0x%02X\n", c))
                in_file:close()
                out_file:close()
                return 1
            end
        end

        out_file:write(string.char(byte))
    end

    in_file:close()
    out_file:close()
    return 0
end

-- Main program
local function main()
    local encode_file = nil
    local decode_file = nil
    local output_file = nil

    local args = arg

    local i = 1
    while i <= #arg do
        local a = arg[i]
        if a == "-e" and i + 1 <= #arg then
            encode_file = arg[i + 1]
            i = i + 2
        elseif a == "-d" and i + 1 <= #arg then
            decode_file = arg[i + 1]
            i = i + 2
        elseif a == "-o" and i + 1 <= #arg then
            output_file = arg[i + 1]
            i = i + 2
        else
            print_help(arg[0] or "utf-256.lua")
            return 1
        end
    end

    if (encode_file and decode_file) or (not encode_file and not decode_file) then
        io.stderr:write("Error: You must specify either -e or -d, but not both.\n")
        print_help(arg[0] or "utf-256.lua")
        return 1
    end

    if not output_file then
        io.stderr:write("Error: Output file is required (-o).\n")
        print_help(arg[0] or "utf-256.lua")
        return 1
    end

    local result = 0
    if encode_file then
        result = encode(encode_file, output_file)
        if result == 0 then
            print("Encoded successfully to '" .. output_file .. "'")
        end
    elseif decode_file then
        result = decode(decode_file, output_file)
        if result == 0 then
            print("Decoded successfully to '" .. output_file .. "'")
        end
    end

    return result
end

os.exit(main())

File: src/c++/utf-256.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>

constexpr size_t CHUNK_SIZE = 8;

void print_help(const std::string &progname) {
    std::cerr << "\nUTF-256 Encoder/Decoder\n\n"
              << "Usage:\n"
              << "  " << progname << " -e UTF8_FILE -o OUTPUT_FILE\n"
              << "  " << progname << " -d UTF256_FILE -o OUTPUT_FILE\n\n";
}

int encode(const std::string &infile, const std::string &outfile) {
    std::ifstream in(infile, std::ios::binary);
    if (!in) {
        std::cerr << "Error opening input file: " << infile << "\n";
        return 1;
    }

    std::ofstream out(outfile, std::ios::binary);
    if (!out) {
        std::cerr << "Error opening output file: " << outfile << "\n";
        return 1;
    }

    char byte;
    while (in.get(byte)) {
        for (int i = 7; i >= 0; --i) {
            uint8_t bit = (byte >> i) & 1;
            uint8_t out_byte = bit ? 0xFF : 0x00;
            out.put(static_cast<char>(out_byte));
        }
    }

    return 0;
}

int decode(const std::string &infile, const std::string &outfile) {
    std::ifstream in(infile, std::ios::binary);
    if (!in) {
        std::cerr << "Error opening input file: " << infile << "\n";
        return 1;
    }

    std::ofstream out(outfile, std::ios::binary);
    if (!out) {
        std::cerr << "Error opening output file: " << outfile << "\n";
        return 1;
    }

    std::vector<uint8_t> chunk(CHUNK_SIZE);
    while (in.read(reinterpret_cast<char *>(chunk.data()), CHUNK_SIZE)) {
        uint8_t byte = 0;

        for (size_t i = 0; i < CHUNK_SIZE; ++i) {
            if (chunk[i] == 0x00) {
                byte <<= 1;
            } else if (chunk[i] == 0xFF) {
                byte = (byte << 1) | 1;
            } else {
                std::cerr << "Invalid byte in UTF-256 file: 0x" 
                          << std::hex << static_cast<int>(chunk[i]) << "\n";
                return 1;
            }
        }

        out.put(static_cast<char>(byte));
    }

    if (!in.eof()) {
        std::cerr << "Error: file is not a multiple of 8 bytes\n";
        return 1;
    }

    return 0;
}

int main(int argc, char *argv[]) {
    std::string encode_file, decode_file, output_file;

    if (argc < 2) {
        print_help(argv[0]);
        return 1;
    }

    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        if (arg == "-e" && i + 1 < argc) {
            encode_file = argv[++i];
        } else if (arg == "-d" && i + 1 < argc) {
            decode_file = argv[++i];
        } else if (arg == "-o" && i + 1 < argc) {
            output_file = argv[++i];
        } else {
            print_help(argv[0]);
            return 1;
        }
    }

    if ((!encode_file.empty() && !decode_file.empty()) ||
        (encode_file.empty() && decode_file.empty())) {
        std::cerr << "Error: You must specify either -e or -d, but not both.\n";
        print_help(argv[0]);
        return 1;
    }

    if (output_file.empty()) {
        std::cerr << "Error: Output file is required (-o).\n";
        print_help(argv[0]);
        return 1;
    }

    int result = 0;

    if (!encode_file.empty()) {
        result = encode(encode_file, output_file);
        if (result == 0) {
            std::cout << "Encoded successfully to '" << output_file << "'\n";
        }
    } else if (!decode_file.empty()) {
        result = decode(decode_file, output_file);
        if (result == 0) {
            std::cout << "Decoded successfully to '" << output_file << "'\n";
        }
    }

    return result;
}
